src/python/WMCore/MicroService/Unified/MSMonitor.py

```diff
@@ -30,52 +42,105 @@ def execute(self, reqStatus):
+                        "+++ request %s all transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
                 # if pileup transfers are completed AND
                 # some input blocks are completed,
                 # move the request status staging -> staged
-                elif self.pileupTransfersCompleted(transferRecords):
+                elif self.completion(transfers, pileup=True):
                     self.logger.debug(
-                        "+++ request %s pileup transfers are completed", req)
-                    self.change(req, 'staged', '+++ monit')
+                        "+++ request %s pileup transfers are completed", reqName)
+                    completedWorkflows.append(wname)
+                    reqStatus[reqName] = 'staged'
                 # transfers not completed
                 # just update the database with their completion
                 else:
                     self.logger.debug(
-                        "+++ request %s transfers are %s completed",
-                        req, completion)
-                    # for us workflow name is the same as request name
-                    wname = reqName
-                    requestStatuses[wname] = transferRecords
-            self.updateTransferInfo(requestStatuses)
+                        "+++ request %s transfers are completed", reqName)
+            self.updateTransferInfo(transferRecords, completedWorkflows)
+            # finally, update statuses for requests
+            for reqName, status in reqStatus.items():
+                self.change(reqName, status, '+++ monit')
         except Exception as err:  # general error
             self.logger.exception('+++ monit error: %s', str(err))
 
-    def getTransferRecords(self, reqName):
+    def completion(self, transferRecords, pileup=None):
+        """
+        Helper function to calculate completion of given records
+        :param records: list of status records
+        :param pileup: check pileup completion (optional)
+        :return: completion status
+        """
+        # check completion of all transfers
+        status = False
+        transferTypes = ['primary', 'secondary']
+        for rec in transferRecords:
+            campaign = rec['campaignName']
+            cdict = self.getCampaignConfig(campaign)
+            thr = cdict.get('PartialCopy', 0)
+            if pileup and rec['dataType'] not in transferTypes:
+                continue
+            if rec['completion'] >= thr:
+                status = True
+        return status
+
+    def getCampaignConfig(self, campaign):
+        "Helper function to get given campaign from backend or cache"
+        if campaign not in self.campaigns:
+            campaigns = self.reqmgrAux.getCampaignConfig("ALL_DOCS")
+            for cdict in campaigns:
+                if cdict not in self.campaigns:
```

Comment:
Testing if a dict is in another dict? If won't work

Reply:
done